File: cs_interface/packetizer.py
#!/usr/bin/python

import serial


class Packetizer():
    def __init__(self, Port, baudrate=115200, start_byte = 0x0A):
        self.start_byte = start_byte
        self.ser = serial.Serial()
        self.ser.baudrate = baudrate

        if(type(Port) is int): #windows numerical port
            self.ser.port = Port - 1 
        elif(type(Port) is str): #linux string port
            self.ser.port = Port

        self.ser.open()

    def get_lock(self):
        current_byte = 0
        packet_array = []
        in_sync = False

        #flush the input buffer to
        #ensure we are reading the latest data
        self.ser.flushInput()
		
        print "Acquiring stream sync..."
        while in_sync == False:
			
            #get a single byte
            current_byte = ord(self.ser.read())

            #if byte was control byte, assume for now we are in sync
            if current_byte == self.start_byte:
                packet_array = [] #clear the array
                packet_array.append(current_byte)

                #read one more byte, should be the size byte
                size = ord(self.ser.read())

                #read <size> more bytes to get the rest of the packet
                self.ser.read(size)
				
                #read another byte, should be control_byte
                current_byte = ord(self.ser.read())

                #if we got a start byte, then we are fairly
                #sure we are in sync
                if (current_byte == self.start_byte):
                    size = ord(self.ser.read())
                    self.ser.read(size)
                    in_sync = True
                    print "sync locked"
					
        #end get_lock()

    def get_packet(self):
        success = False

        while success == False:
            #receive the first two bytes
            #should be start byte and size byte
            header = self.ser.read(2)
				
            #pull out the start and size bytes
            start = ord(header[0])
            size = ord(header[1])

            if (start != self.start_byte): #check that first byte is the start byte
                print "Error: lost sync"
                self.get_lock() #if not, get back into sync
            else: #we are in sync
                if(size > 0): #check that the packet's body isn't empty
                    chr_packet = self.ser.read(size) #read the rest of the packet
                    byte_packet = []
                for character in chr_packet: #convert chars to real bytes
                    byte_packet.append(ord(character))
                success = True

        return byte_packet #return the body of the packet (no header)
	#end get_packet()

	def send_packet(self, data):
		packet = []

		packet.append(chr(self.start_byte)) #add the start byte
		packet.append(chr(len(data)))	   #add the size
		#convert all the data byte into chars
		for byte in data:
			packet.append(chr(byte))

		str_packet = ''.join(packet)

		#send the packet
		self.ser.write(str_packet)
	#end send_packet()

        	

File: cs_interface/cpp/Packetizer.cpp
#include "Packetizer.hpp"

    Packetizer::Packetizer(const char *port_name, uint8_t control_byte)
    {
        this->control_byte = control_byte;
        serial = new Serial(port_name);
    }

    Packetizer::~Packetizer()
    {
        delete serial;
    }

    int Packetizer::get(uint8_t *buf)
    {
        uint8_t header[2];
        bool success = false;
        int size = -1;

        while (!success)
        {
            //get the packet header
            serial->sread(header, 2);

            //check for control byte
            if(header[0] != this->control_byte)
            {
                //we don't have sync, aquire sync
                this->get_lock();
                continue;
            }//else, we are in sync, read the data in the packet


            size = header[1]; //extract the packet size
            serial->sread(buf, header[1]); //read the data into the buffer
            success = true;
        }

        return size;
    }

    int Packetizer::send(uint8_t *buf, uint8_t num)
    {
        uint8_t send_buf[MAX_PACKET_SIZE+2];

        //fill out the header info
        send_buf[0] = this->control_byte;
        send_buf[1] = num;

        //place data into the transmit buffer
        memcpy(send_buf+2, buf, num);

        //send the packet
        serial->swrite(send_buf, num+2);

        return 0;
    }

    void Packetizer::get_lock(void)
    {
        uint8_t buf[MAX_PACKET_SIZE+2];
        uint8_t tmp_byte;
        bool in_sync = false;

        printf("Aquiring stream sync");
        while(!in_sync)
        {
            //get a single byte
            serial->sread(&tmp_byte, 1);

            //if current byte is control byte,
            //assume we're in sync
            if(tmp_byte == this->control_byte)
            {
                //read what should be the size byte
                serial->sread(&tmp_byte, 1);

                //read the rest of the packet(length of size byte),
                //plus two more bytes (should be next packet's control and size bytes)
                serial->sread(buf, tmp_byte+2);

                //check again for new start byte. If it's right,
                //then we can say with a high degree of certainty we
                //are in sync
                if(buf[tmp_byte] == this->control_byte)
                {
                    in_sync = true;
                    //read and throw away the rest of the packet
                    serial->sread(buf, buf[tmp_byte+1]);
                }
            }
        }
        printf("Sync locked!");
    }
